core/forkchoicer.go

```diff
@@ -0,0 +1,127 @@
+	GetTd(common.Hash, uint64) *big.Int
+}
+
+// ForkChoicer is the fork choicer based on the highest total difficulty of the
+// chain(the fork choice used in the eth1) and the external fork choice (the fork
+// choice used in the eth2). This main goal of this ForkChoicer is not only for
+// offering fork choice during the eth1/2 merge phase, but also keep the compatibility
+// for all other proof-of-work networks.
+type ForkChoicer struct {
+	chain ChainReader
+	seed  *mrand.Rand
+
+	// transitioned is the flag whether the chain has finished the
+	// ethash -> transition. It's triggered by receiving the first
+	// "NewBlock" message from the external consensus engine.
+	transitioned uint32
+
+	// preserve is a helper function used in td fork choice.
+	// Miners will prefer to choose the local mined block if the
+	// local td is equal to the extern one. It can nil for light
+	// client
+	preserve func(header *types.Header) bool
+}
+
+func NewForkChoicer(chainReader ChainReader, transitioned bool, preserve func(header *types.Header) bool) *ForkChoicer {
+	// Seed a fast but crypto originating random generator
+	seed, err := crand.Int(crand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		log.Crit("Failed to initialize random seed", "err", err)
+	}
+	forker := &ForkChoicer{
+		chain:    chainReader,
+		seed:     mrand.New(mrand.NewSource(seed.Int64())),
+		preserve: preserve,
+	}
+	if transitioned {
+		forker.SetTransitioned()
+	}
+	return forker
+}
+
+// Reorg returns the result whether the reorg should be applied
+// based on the given external header and local canonical chain.
+// In the td mode, the new head is chosen if the corresponding
+// total difficulty is higher.
+func (f *ForkChoicer) Reorg(header *types.Header) (bool, error) {
+	// If the chain is already transitioned into the casper phase,
+	// always return true because the head is already decided by
+	// the external fork choicer.
+	if f.IsTransitioned() {
+		return true, nil
+	}
+	var (
+		headHeader = f.chain.CurrentHeader()
+		localTD    = f.chain.GetTd(headHeader.Hash(), headHeader.Number.Uint64())
+		externTd   = f.chain.GetTd(header.Hash(), header.Number.Uint64())
+	)
+	if localTD == nil || externTd == nil {
+		return false, errors.New("missing td")
+	}
+	// If the total difficulty is higher than our known, add it to the canonical chain
+	// Second clause in the if statement reduces the vulnerability to selfish mining.
+	// Please refer to http://www.cs.cornell.edu/~ie53/publications/btcProcFC.pdf
+	reorg := externTd.Cmp(localTD) > 0
+	if !reorg && externTd.Cmp(localTD) == 0 {
+		number, headNumber := header.Number.Uint64(), headHeader.Number.Uint64()
+		if number < headNumber {
+			reorg = true
+		} else if number == headNumber {
+			var currentPreserve, externPreserve bool
+			if f.preserve != nil {
+				currentPreserve, externPreserve = f.preserve(headHeader), f.preserve(header)
+			}
+			reorg = !currentPreserve && (externPreserve || mrand.Float64() < 0.5)
```

Comment:
Why do you create your own random source if you're using the global one here? Also, do you need a 64 bit float here? I guess `Intn(2) == 0` is probably cheaper (I'm not totally sure).

Reply:
Yeah, I will fix it by using the own one.

The float64 randomness is implemented long long time ago. I just copied the old TD fork choice, perhaps it's better to keep the original behavior?